File: src/Ranker.cpp
#include "Ranker.h"
#include <limits>
#include <stdexcept>

const int RankerTable::INF = std::numeric_limits<int>::max();

RankerTable::RankerTable(const std::string& text, int alphabetSize)
        : text(text), alphabetSize(alphabetSize) {
    int n = text.size();
    xTable.assign(n + 1, std::vector<int>(alphabetSize, INF));
    yTable.assign(n + 1, std::vector<int>(alphabetSize, -1));

    for (int i = 0; i < alphabetSize; ++i) {
        charToIndexMap['a' + i] = i;
    }
}

int RankerTable::charToIndex(char c) const {
    auto it = charToIndexMap.find(c);
    if (it == charToIndexMap.end())
        throw std::invalid_argument("Character not in alphabet");
    return it->second;
}

void RankerTable::buildXRankerTable() {
    int n = text.size();
    std::vector<int> next(alphabetSize, INF);

    for (int i = n - 1; i >= 0; --i) {
        next[charToIndex(text[i])] = i + 1;
        for (int c = 0; c < alphabetSize; ++c) {
            xTable[i][c] = next[c];
        }
    }
}

void RankerTable::buildYRankerTable() {
    int n = text.size();
    std::vector<int> prev(alphabetSize, -1);

    for (int i = 0; i < n; ++i) {
        prev[charToIndex(text[i])] = i + 1;
        for (int c = 0; c < alphabetSize; ++c) {
            yTable[i + 1][c] = prev[c];
        }
    }
}

int RankerTable::getX(int index, char c) const {
    return xTable[index][charToIndex(c)];
}

int RankerTable::getY(int index, char c) const {
    return yTable[index][charToIndex(c)];
}

File: animations/data/Ranker.py
# data/Ranker.py

class RankerTable:
    INF = float('inf')

    def __init__(self, text: str, alphabet_size: int):
        self.text = text
        self.alphabet_size = alphabet_size
        self.char_to_index_map = {chr(ord('a') + i): i for i in range(alphabet_size)}

        n = len(text)
        self.x_table = [[RankerTable.INF] * alphabet_size for _ in range(n + 1)]
        self.y_table = [[-1] * alphabet_size for _ in range(n + 1)]

    def char_to_index(self, c: str) -> int:
        if c not in self.char_to_index_map:
            raise ValueError(f"Character '{c}' not in alphabet")
        return self.char_to_index_map[c]

    def build_x_ranker_table(self):
        n = len(self.text)
        next_pos = [RankerTable.INF] * self.alphabet_size

        for i in range(n - 1, -1, -1):
            next_pos[self.char_to_index(self.text[i])] = i + 1
            for c in range(self.alphabet_size):
                self.x_table[i][c] = next_pos[c]

    def build_y_ranker_table(self):
        n = len(self.text)
        prev = [-1] * self.alphabet_size

        for i in range(n):
            prev[self.char_to_index(self.text[i])] = i + 1
            for c in range(self.alphabet_size):
                self.y_table[i + 1][c] = prev[c]

    def get_x(self, index: int, c: str) -> int:
        return self.x_table[index][self.char_to_index(c)]

    def get_y(self, index: int, c: str) -> int:
        return self.y_table[index][self.char_to_index(c)]
